coffee-jpa/src/main/java/hu/icellmobilsoft/coffee/jpa/sql/batch/BatchService.java

```diff
@@ -19,873 +19,1062 @@
+        } else if (value instanceof LocalTime) {
+            LocalTime localTime = (LocalTime) value;
+            time = Time.valueOf(localTime);
+        } else if (value instanceof OffsetTime) {
+            OffsetTime offsetTime = ((OffsetTime) value).withOffsetSameInstant(getZoneOffset(ZoneId.systemDefault()));
+            time = Time.valueOf(offsetTime.toLocalTime());
+        } else if (value instanceof Date) {
+            Date date = (Date) value;
+            time = new Time(date.getTime());
+        } else if (value instanceof Calendar) {
+            Calendar calendar = (Calendar) value;
+            time = new Time(calendar.getTimeInMillis());
+        }
+
+        return time;
+    }
+
+    /**
+     * Sets a timestamp object in the prepared statement.<br>
+     * Be careful to use {@link Instant}, not every prepared statement implementation has converter for it.<br>
+     * For example oracle prepared statement throws java.sql.Exception, because there is no converter for java.time.Instant to oracle.sql.TIMESTAMP.
+     * 
+     * @param ps
+     *            the prepared statement.
+     * @param parameterIndex
+     *            index of the parameter in the prepared statement.
+     * @param value
+     *            value of the parameter.
+     * @throws SQLException
+     *             in case of any exception occurs during the process.
+     */
+    protected void setTimestampPsObject(PreparedStatement ps, int parameterIndex, Object value) throws SQLException {
+        Timestamp timestamp = toSQLTimestamp(value);
+        // Ilyennek nem lenne szabad előfordulnia, de ha mégis null a konvertált timestamp, akkor a részünkről nem lefedett típussal van dolgunk,
+        // ezért ilyenkor rábízzuk a driver-re, hogy próbálja meg feloldani.
+        if (timestamp == null) {
+            log.debug("Unhandled timestamp type: [{0}]", value.getClass().getSimpleName());
+            ps.setObject(parameterIndex, value);
+        } else if (dbTimeZone == null) {
+            ps.setTimestamp(parameterIndex, timestamp);
+        } else {
+            ps.setTimestamp(parameterIndex, timestamp, Calendar.getInstance(dbTimeZone));
+        }
+    }
+
+    /**
+     * Converts an object to {@link java.sql.Timestamp}.
+     *
+     * @param value
+     *            the object.
+     * @return the converted {@link java.sql.Timestamp} or null in case of unhandled java type.
+     */
+    protected Timestamp toSQLTimestamp(Object value) {
+        Timestamp timestamp = null;
+
+        if (value instanceof Timestamp) {
+            timestamp = (Timestamp) value;
+        } else if (value instanceof LocalDateTime) {
+            LocalDateTime localDateTime = (LocalDateTime) value;
+            timestamp = Timestamp.valueOf(localDateTime);
+        } else if (value instanceof OffsetDateTime) {
+            LocalDateTime localDateTime = ((OffsetDateTime) value).atZoneSameInstant(ZoneId.systemDefault()).toLocalDateTime();
```

Comment:
Itt biztos jó a systemDefault, ha kezelünk külön dbTimeZone-t is, nem lesz így baj, ha ezek eltérnek? 
pl. ha a systemDefault zone Europe/Budapest és ide bekerül pl 
`2023-02-02T13:06:00Z`, 
abból a zoneSameInstant(defalt).localDateTime után ez a localDateTime  
`2023-02-02T14:06:00`
lesz, ha emellett a dbTimeZone az UTC-re van állítva, akkor valami ilyesmi fut le
`ps.setTimestamp(parameterIndex, Timestamp.valueOf(2023-02-02T14:06:00), Calendar.getInstance("UTC"));`
mi kerül a db-be és mi lesz kiolvasva ha utána selectelünk rajta?
Nem mélyültem el az sql .Timestamp meg a hibernate db time zone kezelésben, úgyhogy lehet jó lesz így, csak érdekel nem lehet e gond, ha a systemDefault zone eltér a hibernate-estől..

Reply:
A hibernate 2 részre osztotta szét most a típusokat, java és jdbc szinten.
Maga a dátum és időtípusokhoz timestamp jdbc típust rendel alapértelmezetten (ezt configgal lehet módosítani, hogy a zone típusok esetén timestamp with timezone legyen, de ezt csak mellékesen jegyzem meg).
A timestamp jdbc típushoz tartozó osztály a `TimestampJdbcType`.
Ha itt megnézed a `doBind()` metódust akkor látható pontosan, hogyan is kezeli a bementést.
Elsőkörben timestamp típusra kerül átalakításra, amit a `javaType.unwrap` végez el.
Maga az OffsetDateTime-hoz tartozó hibernate osztály az `OffsetDateTimeJavaType`, amiben az unwrap-nál látod ugyan ezt a logikát. Annyi eltérsében, hogy 1905 előtti időket kezelik így, mert ott az Instant módszer nem működik. Viszont én azért alkalmaztak csak ezt, mert ez minden 1905 előtt és utáni időben is jó.

Tehát röviden a hibernate minta szerint lett beletéve, tehát ugyan az a módszer.
Amit én követtem logika időzónák terén, hogy minden esetben a rendszer időr állítom át, majd ebből jön a timestamp típus.
Ez azért kell mert a JDBC driver, ha azon setTimestamp metódussal van meghívva, aminek Calendaron keresztül zónát adunk át, akkor ő minden timestamp-et a rendszeridő mentén mér össze az átadott időzónával és aszerint tolja el.

A sampler projekthez készült rengeteg teszt, ez viszont még később kerül majd kiírásra, viszont átírtam a postgresql-hez tartozó dockerfájl időzónáját Europe/Budapest-re és a fenti példádra ezek az eredmények születtek:

OffsetDateTime - 2023-02-02T13:06:00Z 
Timestamp - 2023-02-02 14:06:00.0
Bementett adat - 2023-02-02 13:06:00.000000
Kiolvasott adat - 2023-02-02T14:06 + 01:00
Kiolvasott adat UTC - 2023-02-02T13:06:00Z 

Ha bővebben szeretnéd, ezt megtudjuk nézni élőben is más tesztesetekre is.

Reply:
Köszi, így már világos, hogy jól működik.
Ezt nem tudtam:
> ő minden timestamp-et a rendszeridő mentén mér össze az átadott időzónával és aszerint tolja el.

de így akkor tök jó a megoldás.